File: swivel6.php
<?php

class Swivel6 {
  private $user, $password, $group, $root_url;

  public function __construct($user, $password, $group, $root_url = 'https://api.swivel.com/v1') {
    $this->user = $user;
    $this->password = $password;
    $this->group = $group;
    $this->root_url = $root_url;
  }

  public function charts() {
    return $this->api_get('groups/'.$this->group.'/charts');
  }

  public function create_chart() {
    return $this->api_post('groups/'.$this->group.'/charts');
  }

  public function update_chart($chart, $data) {
    $this->api_put('groups/'.$this->group.'/charts/'.$chart->id, $data);
  }

  public function get_chart_data($chart) {
    return $this->parse_csv($this->req('get','groups/'.$this->group.'/charts/'.$chart->id.'.csv'));
  }

  public function set_chart_data($chart, $data) {
    $this->update_chart($chart, array('data' => $this->create_csv($data)));
  }

  public function append_chart_data($chart, $data) {
    $this->update_chart($chart, array('data' => $this->create_csv($data), 'mode' => 'append'));
  }

  private function api_get($url, $params = false) {
    return $this->xml2object($this->req('get', $url.'.xml', $params));
  }

  private function api_post($url, $params = false) {
    return $this->xml2object($this->req('post', $url.'.xml', $params));
  }

  private function api_put($url, $params = false) {
    $this->req('put', $url.'.xml', $params);
  }

  private function req($type, $url, $params = false) {
    $ch = curl_init($this->root_url.'/'.$url);
    curl_setopt($ch, CURLOPT_VERBOSE, 1);
    curl_setopt($ch, CURLOPT_RETURNTRANSFER, 1);
    curl_setopt($ch, CURLOPT_SSL_VERIFYHOST, 0);
    curl_setopt($ch, CURLOPT_SSL_VERIFYPEER, 0);
    curl_setopt($ch, CURLOPT_HTTPAUTH, CURLAUTH_BASIC);
    curl_setopt($ch, CURLOPT_USERPWD, $this->user.':'.$this->password);
    if($type != 'get') {
      curl_setopt($ch, CURLOPT_CUSTOMREQUEST, strtoupper($type));

      if($params) {
        curl_setopt($ch, CURLOPT_POSTFIELDS, $params);
      }
    }

    $resp = curl_exec($ch);
    curl_close($ch);
    echo $resp;
    return $resp;
  }

  private function create_csv($data) {
    $csv = fopen('php://temp/maxmemory:'. (5*1024*1024), 'r+');
    foreach($data as $line) {
      fputcsv($csv, $line);
    }
    rewind($csv);
    return stream_get_contents($csv);
  }

  private function parse_csv($data) {
    $csv = fopen('php://temp/maxmemory:'. (5*1024*1024), 'r+');
    fwrite($csv, $data);
    rewind($csv);
    $ar = array();
    while(($line = fgetcsv($csv)) !== FALSE) {
      $ar[] = $line;
    }
    return $ar;
  }

  # http://us3.php.net/manual/en/function.xml-parse.php#87920
  private function xml2array($contents, $get_attributes = 1, $priority = 'tag') {
    $parser = xml_parser_create('');
    xml_parser_set_option($parser, XML_OPTION_TARGET_ENCODING, "UTF-8");
    xml_parser_set_option($parser, XML_OPTION_CASE_FOLDING, 0);
    xml_parser_set_option($parser, XML_OPTION_SKIP_WHITE, 1);
    xml_parse_into_struct($parser, trim($contents), $xml_values);
    xml_parser_free($parser);
    if (!$xml_values)
      return; //Hmm...
    $xml_array = array ();
    $parents = array ();
    $opened_tags = array ();
    $arr = array ();
    $current = & $xml_array;
    $repeated_tag_index = array ();
    foreach ($xml_values as $data)
    {
      unset ($attributes, $value);
      extract($data);
      $result = array ();
      $attributes_data = array ();
      if (isset ($value))
      {
        if ($priority == 'tag')
          $result = $value;
        else
          $result['value'] = $value;
      }
      if (isset ($attributes) and $get_attributes)
      {
        foreach ($attributes as $attr => $val)
        {
          if ($priority == 'tag')
            $attributes_data[$attr] = $val;
          else
            $result['attr'][$attr] = $val; //Set all the attributes in a array called 'attr'
        }
      }
      if ($type == "open")
      {
        $parent[$level -1] = & $current;
        if (!is_array($current) or (!in_array($tag, array_keys($current))))
        {
          $current[$tag] = $result;
          if ($attributes_data)
            $current[$tag . '_attr'] = $attributes_data;
          $repeated_tag_index[$tag . '_' . $level] = 1;
          $current = & $current[$tag];
        }
        else
        {
          if (isset ($current[$tag][0]))
          {
            $current[$tag][$repeated_tag_index[$tag . '_' . $level]] = $result;
            $repeated_tag_index[$tag . '_' . $level]++;
          }
          else
          {
            $current[$tag] = array (
              $current[$tag],
              $result
            );
            $repeated_tag_index[$tag . '_' . $level] = 2;
            if (isset ($current[$tag . '_attr']))
            {
              $current[$tag]['0_attr'] = $current[$tag . '_attr'];
              unset ($current[$tag . '_attr']);
            }
          }
          $last_item_index = $repeated_tag_index[$tag . '_' . $level] - 1;
          $current = & $current[$tag][$last_item_index];
        }
      }
      elseif ($type == "complete")
      {
        if (!isset ($current[$tag]))
        {
          $current[$tag] = $result;
          $repeated_tag_index[$tag . '_' . $level] = 1;
          if ($priority == 'tag' and $attributes_data)
            $current[$tag . '_attr'] = $attributes_data;
        }
        else
        {
          if (isset ($current[$tag][0]) and is_array($current[$tag]))
          {
            $current[$tag][$repeated_tag_index[$tag . '_' . $level]] = $result;
            if ($priority == 'tag' and $get_attributes and $attributes_data)
            {
              $current[$tag][$repeated_tag_index[$tag . '_' . $level] . '_attr'] = $attributes_data;
            }
            $repeated_tag_index[$tag . '_' . $level]++;
          }
          else
          {
            $current[$tag] = array (
              $current[$tag],
              $result
            );
            $repeated_tag_index[$tag . '_' . $level] = 1;
            if ($priority == 'tag' and $get_attributes)
            {
              if (isset ($current[$tag . '_attr']))
              {
                $current[$tag]['0_attr'] = $current[$tag . '_attr'];
                unset ($current[$tag . '_attr']);
              }
              if ($attributes_data)
              {
                $current[$tag][$repeated_tag_index[$tag . '_' . $level] . '_attr'] = $attributes_data;
              }
            }
            $repeated_tag_index[$tag . '_' . $level]++; //0 and 1 index is already taken
          }
        }
      }
      elseif ($type == 'close')
      {
        $current = & $parent[$level -1];
      }
    }
    return ($xml_array);
  }

  private function array2object(&$ar) {
    $numeric_array = true;
    foreach($ar as $k => $v) {
      if(is_array($v)) $ar[$k] = $this->array2object($v);
      if(!is_numeric($k)) $numeric_array = false;
    }
    $vals = array_values($ar);

    if(count($ar) == 1 and is_array($vals[0])) {
      return array_shift($ar);
    } else if($numeric_array) {
      return $ar;
    } else {
     return (object) $ar;
    }
  }

  private function xml2object($contents) {
    $ar = $this->xml2array($contents, false);
    $root = array_shift($ar);
    return $this->array2object($root);
  }

}

$s6 = new Swivel6('email@address.com', 'YOUR_PASSWORD', 1000000); // NOTE replace 1000000 with your group number

$newchart = $s6->create_chart();
echo $newchart->name."\n";

$s6->update_chart($newchart, array('title' => 'API Chart', 'name' => 'API Chart', 'description' => 'A chart made automatically with the Swivel API.'));
$s6->set_chart_data($newchart, array(
  array('Monthly Summary','New Topics','New Posts','New Members','Most Online'),
  array('March 2009',922,12133,193,26),
  array('February 2009',888,9061,192,36),
  array('January 2009',689,8654,390,21),
  array('December 2008',408,3375,166,14),
  array('November 2008',5,28,0,3),
  array('October 2008',0,0,2,1),
  array('September 2008',2,2,0,1),
  array('August 2008',14,21,9,2),
  array('July 2008',4,9,6,4)));
$s6->append_chart_data($newchart, array(array('April 2009',271,3372,43,25)));

print_r($s6->get_chart_data($newchart));
echo "Created chart at https://business.swivel.com/charts/".$newchart->id."\n";

$charts = $s6->charts();

foreach($charts as $chart) {
  echo $chart->name.": https://business.swivel.com/charts/".$chart->id."\n";
}

?>

File: swivel6.rb
require 'xmlsimple'
require 'uri'
require 'net/https'
require 'json'

class Swivel6
  def initialize user, password, group, root_url = 'https://api.swivel.com/v1'
    @user = user
    @password = password
    @group = group
    @root_url = root_url
    @uri = URI.parse(root_url)
    @http = Net::HTTP.new @uri.host, @uri.port
    if @uri.scheme == 'https'
      @http.use_ssl = true
      @http.verify_mode = OpenSSL::SSL::VERIFY_NONE
    end
  end

  def workbooks
    api_get "groups/#{@group}/workbooks"
  end

  def workbook_sheets workbook
    api_get "workbooks/#{workbook['id']}/sheets"
  end

  def sheet_data sheet
    FasterCSV.parse get("sheets/#{sheet['id']}.csv")
  end

  def charts
    api_get "groups/#{@group}/charts"
  end

  def chart_data chart
    FasterCSV.parse get("charts/#{chart['id']}.csv")
  end

  def create_chart
    api_post "groups/#{@group}/charts", 'chart' => { 'name' => 'API Created Chart' }
  end

  def update_chart chart, hash
    api_put "groups/#{@group}/charts/#{chart['id']}", 'chart' => hash
  end

  def clear_chart_data chart
    api_put "charts/#{chart['id']}", 'chart' => { 'data' => '' }
  end

  def set_chart_data chart, data
    api_put "charts/#{chart['id']}", 'chart' => { 'data' => data.map{|row| row.to_csv}.join }
  end

private
  def api_post path, hash = {}
    xml = XmlSimple.xml_out hash, 'keeproot' => true, 'noattr' => true
    resp = post(path + ".json", xml).body
    JSON.parse resp
  end

  def api_put path, hash = {}
    xml = XmlSimple.xml_out hash, 'keeproot' => true, 'noattr' => true
    resp = put(path + ".json", xml).body
    JSON.parse resp if resp.strip != ''
  end

  def api_get path, params = {}
    ret = []
    json = nil
    page = 1
    max_page = 1_000
    begin
      resp = get path + ".json", params.merge(:page => page)

      begin
        json = JSON.parse resp.body
      rescue
        puts path
        puts resp.body
      end

      return json unless json.is_a? Array

      ret.concat json

      should_continue =
        if resp['content-range'] && resp['content-range'] =~ /pages \d+-(\d+)\/(\d+)/
          page = $1.to_i
          max_page = $2.to_i
          page < max_page
        else
          # hack for pages that don't have content-range
          json.length == 30
        end

      page += 1
    end while should_continue
    ret
  end

  def get path, params = {}
    request 'get', path, :params => params
  end

  def post path, xml
    request 'post', path, :xml => xml
  end

  def put path, xml
    request 'put', path, :xml => xml
  end

  def request req, path, opts = {}
    uri = @uri.merge path
    uri.query = hash_to_param_string(opts[:params]) if opts[:params]

    puts "#{req} #{uri.request_uri}"
    req_class = case req
      when "post"
        Net::HTTP::Post
      when "put"
        Net::HTTP::Put
      when "get"
        Net::HTTP::Get
      end
    req = req_class.new uri.request_uri

    if opts[:xml]
      req.set_content_type 'text/xml'
      req.body = opts[:xml]
    end

    req.basic_auth @user, @password

    @http.start do |http|
      http.request req
    end
  end

  def hash_to_param_string hash
    hash.keys.map{ |k| CGI.escape(k.to_s) + '=' + CGI.escape(hash[k].to_s) }.join('&');
  end
end